### PR TITLE
feature(azure): basic provisioner for azure

### PR DESCRIPTION
**Pull request by soyacz**
Initial version of azure scylla provider.
It's missing many features - will be added when needed.
For now it can create VM from scylla image.

## PR pre-checks (self review)

- [ ] I followed [KISS principle](https://en.wikipedia.org/wiki/KISS_principle) and [best practices](https://docs.google.com/document/d/1jihgOKb5iGRlD8_HQ92O0JbLk1kASUoZT23i_MXFSKI)
- [ ] I didn't leave commented-out/debugging code
- [ ] I added the relevant `backport` labels
- [ ] New configuration option are added and documented (in `sdcm/sct_config.py`)
- [ ] I have added tests to cover my changes (Infrastructure only - under `unit-test/` folder)
- [ ] All new and existing unit tests passed (CI)
- [ ] I have updated the Readme/doc folder accordingly (if needed)

### CODE REVIEW

**Comment by dkropachev** [comment]
@soyacz , I know that it is a draft, but I mind you that aws provisioner and azure should share same logic.

**Comment by soyacz** [comment]
> @soyacz , I know that it is a draft, but I mind you that aws provisioner and azure should share same logic.

In future - yes. I saw your solution for aws and draft for azure - I see some problems with it. In long term will be very costly:
1. No separation of logic from sct
2. No separation from plan of whole test env creation
2. many copy-paste code (aws->azure)
3. very, very complicated

Idea of this PR is to have Azure provisioner with very simple api designed for Scylla testing (in general, not bound to SCT). Later, we'll create provisioners with the same api for aws and gce. 
Having such provisioner we can design logic for creating whole test env separately from provisioner (e.g. using Strategy Design Pattern). 
I also disagree with sharing logic - In my opinion we should rather share API.

**Comment by dkropachev** [comment]
> > @soyacz , I know that it is a draft, but I mind you that aws provisioner and azure should share same logic.
> 
> In future - yes. I saw your solution for aws and draft for azure - I see some problems with it. In long term will be very costly:
> 
> 1. No separation of logic from sct
> 2. No separation from plan of whole test env creation
> 3. many copy-paste code (aws->azure)
> 4. very, very complicated
> 
> Idea of this PR is to have Azure provisioner with very simple api designed for Scylla testing (in general, not bound to SCT). Later, we'll create provisioners with the same api for aws and gce. Having such provisioner we can design logic for creating whole test env separately from provisioner (e.g. using Strategy Design Pattern). I also disagree with sharing logic - In my opinion we should rather share API.

Current provisioner is not perfect in any way, but I see that you not quite get it, here is details that could be helpful if you will be willing to look at it:
1. Common provisioner logic is located at `sdcm/provisioner`, while everything that is linked to sct ls located `sdcm/sct_provisioner`.
2. The only link between sct and provisioner is `SCTConfiguration`, beside it there is `TestConfig` which is temporarily needed to be integratable into sct it self.
3. `SCTConfiguration` is just a configuration format/language you anyway will need to have, picking it up from sct, is just a smart choice.

Having that in mind you above statements are not correct:
> 1. No separation of logic from sct
1. Not true. Beside that I see that your current provision logic is already tight to sct, via `InstanceDefinition.purpose` at `create_virtual_machine`.
> 2. No separation from plan of whole test env creation
2. Not true, you can deploy any part of test env separately
> 3. many copy-paste code (aws->azure)
3. Some parts should be moved to `common` folder, and be reused, so of it should
> 4. very, very complicated
4. Any solution that have abstraction levels will be complicated. Abstraction level will be needed when you will start cover more than one cloud provider. I see that you right now under the spell of thinking that it could be done simpler, but it is just because you are locked on one particular cloud provider and one simple configuration ignoring all possible configurations such as multiple interfaces, multiple disks of different types, multiple DCs hosting on different regions, and pricing.

I understand that you don't want to look at the current provider, but I mind you to look at `old aws provision logic` located at `cluster_aws.py` to have a idea of all cases you will need to cover and get better in AWS provision logic.

**Comment by soyacz** [comment]
> Current provisioner is not perfect in any way, but I see that you not quite get it, here is details that could be helpful if you will be willing to look at it:
> 
> 1. Common provisioner logic is located at `sdcm/provisioner`, while everything that is linked to sct ls located `sdcm/sct_provisioner`.
> 2. The only link between sct and provisioner is `SCTConfiguration`, beside it there is `TestConfig` which is temporarily needed to be integratable into sct it self.

While there's `sdcm/provisioner/common/provisioner.py` with abstract class it cannot serve as common interface for creating instances. It is due `InstanceParams` attribute which is different for each provider. This forces provisioning to be aware what provider is used. This leads to leaking provisioner logic to higher levels - in this way SCT itself (as it's the only client).

My code is about creating provisioner that does not require information of what kind it is because uses always the same input parameters. So when you want to create VM, client of this api must just tell: name, size, tags, disk sizes etc. (my current implementation does not cover everything yet) regardless of cloud provider (params are cloud agnostic).

> 3. `SCTConfiguration` is just a configuration format/language you anyway will need to have, picking it up from sct, is just a smart choice.
> 
> Having that in mind you above statements are not correct:
> 
> > 1. No separation of logic from sct
> 
> 1. Not true. Beside that I see that your current provision logic is already tight to sct, via `InstanceDefinition.purpose` at `create_virtual_machine`.
As stated above - using provisioner requires knowledge what type of provisioner it is - in this terms it leaks to sct.

`purpose` is not related to SCT - it's about providing information what's the goal of creating this instance. This provisioner is suited purely for Scylla testing and some information about role in tests should be provided. It is used for filtering (e.g. when we discover existing vm's - e.g. when reuse cluster - we need this information).
> 
> > 2. No separation from plan of whole test env creation
> 
> 2. Not true, you can deploy any part of test env separately
> 
Again, plan must be aware what kind of provisioner uses.
> > 3. many copy-paste code (aws->azure)
> 
> 3. Some parts should be moved to `common` folder, and be reused, so of it should
> 
In terms of instance provisioning, there's not much in common - each cloud provider has own api, own ways of naming things, own flow. Later part, e.g. provisioning set of vm's for SCT, should require just one class that does it and takes provisioner as parameter (regardless what kind of provisioner it is, it should just work). Generally I try to use composition-over-inheritance programming paradigm - which is less complex and provides more flexibility.

> > 4. very, very complicated
> 
> 4. Any solution that have abstraction levels will be complicated. Abstraction level will be needed when you will start cover more than one cloud provider. I see that you right now under the spell of thinking that it could be done simpler, but it is just because you are locked on one particular cloud provider and one simple configuration ignoring all possible configurations such as multiple interfaces, multiple disks of different types, multiple DCs hosting on different regions, and pricing.
> 
I think we don't need so many abstraction levels. When I think as provider client who just wants provision scylla node, I think of it name, size, region, version, disk size etc. - not in terms of what classes of each component are and how to use it.

Yes, now it misses many functions, which will be added - but anyway, there's no need for multiple abstraction layers.


> I understand that you don't want to look at the current provider, but I mind you to look at `old aws provision logic` located at `cluster_aws.py` to have a idea of all cases you will need to cover and get better in AWS provision logic.

Is there any particular thing that cannot be implemented by providing a param to `provider` that is not common to all cloud providers?

**Comment by dkropachev** [comment]
> > Current provisioner is not perfect in any way, but I see that you not quite get it, here is details that could be helpful if you will be willing to look at it:
> > 
> > 1. Common provisioner logic is located at `sdcm/provisioner`, while everything that is linked to sct ls located `sdcm/sct_provisioner`.
> > 2. The only link between sct and provisioner is `SCTConfiguration`, beside it there is `TestConfig` which is temporarily needed to be integratable into sct it self.
> 
> While there's `sdcm/provisioner/common/provisioner.py` with abstract class it cannot serve as common interface for creating instances. It is due `InstanceParams` attribute which is different for each provider. This forces provisioning to be aware what provider is used. 
You can easily solve it, by consuming dict from `AttrBuilder` or taking it's instance instead. 

> This leads to leaking provisioner logic to higher levels - in this way SCT itself (as it's the only client).
Not true, provision logic is not in any shape of form in SCT, just run grep to see it.

> My code is about creating provisioner that does not require information of what kind it is because uses always the same input parameters. So when you want to create VM, client of this api must just tell: name, size, tags, disk sizes etc. (my current implementation does not cover everything yet) regardless of cloud provider (params are cloud agnostic).
It is exactly what you get by moving `AttrBuilder` into `InstanceProvisionerBase.provision` parameters. 

> 
> > 3. `SCTConfiguration` is just a configuration format/language you anyway will need to have, picking it up from sct, is just a smart choice.
> > 
> > Having that in mind you above statements are not correct:
> > > 1. No separation of logic from sct
> > 
> > 
> > 
> > 1. Not true. Beside that I see that your current provision logic is already tight to sct, via `InstanceDefinition.purpose` at `create_virtual_machine`.
> >    As stated above - using provisioner requires knowledge what type of provisioner it is - in this terms it leaks to sct.
> 
> `purpose` is not related to SCT - it's about providing information what's the goal of creating this instance. This provisioner is suited purely for Scylla testing and some information about role in tests should be provided. It is used for filtering (e.g. when we discover existing vm's - e.g. when reuse cluster - we need this information).
Not true, totaly related, put it outside of SCT box, and it have no meaning, since it represents SCT objects.
Also if you need another type of instance, i.e. purpose, you do what ? Right, you go and change provisioner logic to accomodate it, it is exactly what is called coupling. 

> 
> > > 2. No separation from plan of whole test env creation
> > 
> > 
> > 
> > 2. Not true, you can deploy any part of test env separately
> 
> Again, plan must be aware what kind of provisioner uses.
Nope, it is not, it gets instance parameters, gets provisioner feed instance parameters to the provisioner and done.

> 
> > > 3. many copy-paste code (aws->azure)
> > 
> > 
> > 
> > 3. Some parts should be moved to `common` folder, and be reused, so of it should
> 
> In terms of instance provisioning, there's not much in common - each cloud provider has own api, own ways of naming things, own flow. Later part, e.g. provisioning set of vm's for SCT, should require just one class that does it and takes provisioner as parameter (regardless what kind of provisioner it is, it should just work). Generally I try to use composition-over-inheritance programming paradigm - which is less complex and provides more flexibility.
> 
> > > 4. very, very complicated
> > 
> > 
> > 
> > 4. Any solution that have abstraction levels will be complicated. Abstraction level will be needed when you will start cover more than one cloud provider. I see that you right now under the spell of thinking that it could be done simpler, but it is just because you are locked on one particular cloud provider and one simple configuration ignoring all possible configurations such as multiple interfaces, multiple disks of different types, multiple DCs hosting on different regions, and pricing.
> 
> I think we don't need so many abstraction levels. When I think as provider client who just wants provision scylla node, I think of it name, size, region, version, disk size etc. - not in terms of what classes of each component are and how to use it.
> 
> Yes, now it misses many functions, which will be added - but anyway, there's no need for multiple abstraction layers.
> 
> > I understand that you don't want to look at the current provider, but I mind you to look at `old aws provision logic` located at `cluster_aws.py` to have a idea of all cases you will need to cover and get better in AWS provision logic.
> 
> Is there any particular thing that cannot be implemented by providing a param to `provider` that is not common to all cloud providers?

That is exactly what is done in current provisioner, don't see point here.
My understanding is that number of cases would grow amount of code, once it will become big enought you will split it, and at the end you will endup with same complexity level.

File: sdcm/provision/provisioner.py
Comment (bentsi) [inline sdcm/provision/provisioner.py:39]: user_name

File: sdcm/provision/provisioner.py
Comment (bentsi) [inline sdcm/provision/provisioner.py:40]: ssh_public_key

File: sdcm/provision/provisioner.py
Comment (bentsi) [inline sdcm/provision/provisioner.py:43]: secondory_network_interface

File: sdcm/provision/provisioner.py
Comment (bentsi) [inline sdcm/provision/provisioner.py:44]: add extra_disks (in AWS: EBS)

File: sdcm/provision/provisioner.py
Comment (soyacz) [reply sdcm/provision/provisioner.py:43]: I'll add it when needed

**Comment by soyacz** [comment]
@bentsi can you review it before it gets far too big? And I'm also curious of your insights. Thanks!

**Comment by roydahan** [comment]
@soyacz let's try to do a code walkthrough in this week QA meeting.
Even if it's not ready, it will help getting more people into this code and help with review.

**Comment by soyacz** [comment]
@bentsi @roydahan @enaydanov @yarongilor 
We agreed with Fabio to freeze this branch - please do a review and merge it before doing next steps.

Generally it works - 'topology_changes' nemesis basically work (there are still some errors, but rather not related to provisioning), 3h-longevity also can be run. 

In next PR's we'll be resolving issues one by one.

**Comment by roydahan** [comment]
Thanks @soyacz 
(This was actually my request, I wanted to put a line where we can claim that provisioning and basic functionality works.
Later on, we can do the complete testing for all nemesis and several kind of longevities (and split the work among the team).

**Comment by soyacz** [comment]
I've created fake AzureService that enables fast testing of Azure provisioner: https://github.com/soyacz/scylla-cluster-tests/pull/1
You can check it out, or leave - once current branch is merged I'll reconfigure this PR to use upstream as a destination.

**Comment by roydahan** [comment]
@scylladb/qa-maintainers Please review, you'll all need to be familiar with this code.

**Comment by fgelcer** [comment]
@soyacz , please resolve conflicts

File: defaults/azure_config.yaml
Comment (enaydanov) [inline defaults/azure_config.yaml:14]: Latest CentOS 7 by Rogue Wave Software (formerly OpenLogic) is 7.9

Is it intended to stick with 7.5?

File: vars/getJenkinsLabels.groovy
Comment (enaydanov) [inline vars/getJenkinsLabels.groovy:21]: We have `azure-sct-builders` label and 1 builder is deployed already: https://jenkins.scylladb.com/computer/azure%2Dqa%2Dbuilder1/

File: test-cases/longevity/longevity-azure-nemesis-test.yaml
Comment (enaydanov) [inline test-cases/longevity/longevity-azure-nemesis-test.yaml:21]: Is there a reason to alter ssh_transport for this job?

File: sdcm/provision/scylla_yaml/auxiliaries.py
Comment (enaydanov) [inline sdcm/provision/scylla_yaml/auxiliaries.py:107]: Please, add comma here.

File: sdcm/cluster_azure.py
Comment (enaydanov) [inline sdcm/cluster_azure.py:183]: Please, use a key for GCE instead. We already do this for Azure' SCT runner.

File: sdcm/sct_runner.py
Comment (enaydanov) [inline sdcm/sct_runner.py:817]: It's OK to have no a  public key here?

File: sdcm/tester.py
Comment (enaydanov) [inline sdcm/tester.py:875]: Please, use `scylla-test` key instead

**Review by enaydanov — CHANGES_REQUESTED** [review]

File: sdcm/provision/azure/network_interface_provider.py
Comment (yarongilor) [inline sdcm/provision/azure/network_interface_provider.py:73]: probably more recommended to use the lazy `%s` format in logger messages.

File: sdcm/cluster_azure.py
Comment (fgelcer) [reply sdcm/cluster_azure.py:3]: not sure about it, as new branches (for new versions) will use Python3.10+

File: vars/getJenkinsLabels.groovy
Comment (fgelcer) [reply vars/getJenkinsLabels.groovy:21]: fixed in a separated PR, as we wanted to freeze this PR to get it merged:
https://github.com/scylladb/scylla-cluster-tests/pull/4506/commits/1f3ed9ea0f1efb9d9d94712103c10da655b5cc0f

File: sdcm/cluster.py
Comment (soyacz) [reply sdcm/cluster.py:3907]: should be removed.

File: sdcm/cluster_azure.py
Comment (soyacz) [reply sdcm/cluster_azure.py:3]: 3.10+ should work too.

File: defaults/azure_config.yaml
Comment (soyacz) [reply defaults/azure_config.yaml:2]: anyway SpotTermination thread raises critical event ending the test. Currently we won't detect it so test will fail due to missing nodes or loader error. I'd keep spot to minimize costs.

File: defaults/azure_config.yaml
Comment (soyacz) [reply defaults/azure_config.yaml:14]: it's not intended. Updating it to 7.9

File: vars/getJenkinsLabels.groovy
Comment (soyacz) [reply vars/getJenkinsLabels.groovy:21]: I think It can be done here, along with other small fixes.

File: test-cases/longevity/longevity-azure-nemesis-test.yaml
Comment (soyacz) [reply test-cases/longevity/longevity-azure-nemesis-test.yaml:21]: no, it was part of testing. Removing it.

File: sdcm/sct_runner.py
Comment (soyacz) [reply sdcm/sct_runner.py:817]: For sct runner we use predefined (specialized) image - which already has ssh keys imported (also user prepared).

File: sdcm/provision/azure/network_interface_provider.py
Comment (soyacz) [reply sdcm/provision/azure/network_interface_provider.py:73]: you're right - but I've already did it in this way all over the place - do you think it is worth to spend time on refactoring it?

File: vars/getJenkinsLabels.groovy
Comment (fgelcer) [reply vars/getJenkinsLabels.groovy]: no objections

File: sdcm/provision/azure/network_interface_provider.py
Comment (fgelcer) [reply sdcm/provision/azure/network_interface_provider.py:73]: well, IIRC it doesn't offend lint anymore, so not a must IMHO

Comment (enaydanov) [reply sdcm/provision/azure/network_interface_provider.py:73]: It's worth to convert it to lazy-logging.

Comment (soyacz) [reply sdcm/provision/azure/network_interface_provider.py]: Ok, done.

**Comment by soyacz** [comment]
@enaydanov @yarongilor - thanks for your feedback. I adapted the code to it - is there anything else I could improve?

**Comment by fgelcer** [comment]
ping @bentsi @enaydanov @roydahan

**Review by enaydanov — APPROVED** [review]

File: sdcm/provision/provisioner.py
Comment (fgelcer) [reply sdcm/provision/provisioner.py:43]: @soyacz ?

**Review by fgelcer — APPROVED** [review]
LGTM

**Comment by fgelcer** [comment]
ping @bentsi @roydahan

File: test-cases/longevity/longevity-azure-nemesis-test.yaml
Comment (bentsi) [inline test-cases/longevity/longevity-azure-nemesis-test.yaml:1]: not needed, let's remove

File: jenkins-pipelines/longevity-nemesis-test-azure.jenkinsfile
Comment (bentsi) [inline jenkins-pipelines/longevity-nemesis-test-azure.jenkinsfile:1]: remove

File: test-cases/artifacts/azure-image.yaml
Comment (bentsi) [inline test-cases/artifacts/azure-image.yaml:11]: set to false, keep the todo.

File: test-cases/artifacts/azure-image.yaml
Comment (bentsi) [inline test-cases/artifacts/azure-image.yaml:9]: add todo: need more research

File: test-cases/artifacts/azure-image.yaml
Comment (bentsi) [inline test-cases/artifacts/azure-image.yaml:17]: let's remove (default should be used)

File: unit_tests/provisioner/test_azure_provisioner.py
Comment (bentsi) [inline unit_tests/provisioner/test_azure_provisioner.py:1]: let's disable the test since it creates real resources and re-enable when the mock PR will be ready

**Review by bentsi — APPROVED** [review]

**Pull request merged**
